### PR TITLE
Add chartconfig framework

### PR DESCRIPTION
**Pull request by rossf7**
Towards giantswarm/giantswarm#1902

Adds the chartconfig framework and the chart resource. The resource will be implemented in later PRs. Needs the chartconfig CRD to be merged first in giantswarm/apiextensions/pull/53.

### CODE REVIEW

File: Gopkg.toml [hunk 1]
```diff
@@ -29,6 +29,10 @@ required = [
   "k8s.io/apimachinery/pkg/apis/meta/v1",
 ]
 
+[[constraint]]
+  branch = "add-chartconfig"
```
Comment (rossf7) [inline Gopkg.toml]: TODO

File: service/chartconfig/v1/resource/chart/create.go [hunk 2]
```diff
@@ -0,0 +1,7 @@
+package chart
+
+import "context"
+
+func (r *Resource) ApplyCreateChange(ctx context.Context, obj, createChange interface{}) error {
+	return nil
+}
```
Comment (rossf7) [inline service/chartconfig/v1/resource/chart/create.go]: Chart resource will be implemented in later PRs.

File: service/chartconfig/v1/resource_set.go [hunk 3]
```diff
@@ -0,0 +1,119 @@
...
+	}
+
+	handlesFunc := func(obj interface{}) bool {
+		_, err := key.ToCustomObject(obj)
+		if err != nil {
+			return false
+		}
+
+		// Currently there is only one version to be handled. As long as the
+		// object is of right type, it's good to go.
```
Comment (rossf7) [inline service/chartconfig/v1/resource_set.go]: Versioning will come later.

File: service/chartconfig/v1/key/key.go [hunk 4]
```diff
@@ -0,0 +1,22 @@
...
+// ToCustomObject converts value to v1alpha1.ChartConfig and returns it or error
+// if type does not match.
+func ToCustomObject(v interface{}) (v1alpha1.ChartConfig, error) {
+	customObjectPointer, ok := v.(*v1alpha1.ChartConfig)
+	if !ok {
+		return v1alpha1.ChartConfig{}, microerror.Maskf(wrongTypeError, "expected '%T', got '%T'", &v1alpha1.ChartConfig{}, v)
+	}
+
+	if customObjectPointer == nil {
+		return v1alpha1.ChartConfig{}, microerror.Maskf(emptyValueError,
```
Comment (xh3b4sd) [inline service/chartconfig/v1/key/key.go]: IMO this wrapping with errors does more harm than good.

File: service/chartconfig/v1/resource_set.go [hunk 5]
```diff
@@ -0,0 +1,119 @@
...
+
+	if config.K8sClient == nil {
+		return nil, microerror.Maskf(invalidConfigError, "config.K8sClient must not be empty")
+	}
+	if config.Logger == nil {
+		return nil, microerror.Maskf(invalidConfigError, "config.Logger must not be empty")
+	}
+	if config.ProjectName == "" {
+		return nil, microerror.Maskf(invalidConfigError, "config.ProjectName must not be empty")
+	}
```
Comment (xh3b4sd) [inline service/chartconfig/v1/resource_set.go]: Dependencies and setting have usually their own blocks.

File: service/chartconfig/v1/resource_set.go [hunk 6]
```diff
@@ -0,0 +1,119 @@
...
+
+		metricsWrapConfig := metricsresource.WrapConfig{}
+
+		metricsWrapConfig.Name = config.ProjectName
+
+		resources, err = metricsresource.Wrap(resources, metricsWrapConfig)
+		if err != nil {
+			return nil, microerror.Mask(err)
+		}
+	}
```
Comment (xh3b4sd) [inline service/chartconfig/v1/resource_set.go]: Please check https://github.com/giantswarm/aws-operator/blob/cf464161b1285735857b855b4ff59bf1cf137307/service/awsconfig/v7/cluster_resource_set.go#L293-L326. I would like to use the more scoped version. Then we also do not have to name the config variables but can simply role with `c` and the separated blocks are more readable.

**Comment by rossf7** [comment]
Thanks, the resource wrapping is much nicer that way.

@xh3b4sd Changes made, PTAL.

**Pull request merged**